### ISSUE TITLE
Voting Classifier

### ISSUE THREAD

**Issue by abhikjha** [ageron/handson-ml#210]
Hi Aurelien,

For my new project which is again a binary classification task, I used two models - RandomForest and GradientBoostingClassifer.

RF is predicting class 1 mostly correctly on test set while GradientBoosting is predicting class 0 mostly correctly. I wanted my model to use the best of both worlds. In search of this, I came across Voting Classifier under sklearn which sorts of doing the same thing.

My question is - how do I use this Voting Classifier? Should I chose voting option as "hard" or "soft" with some weights. If later, then how do I choose weights? Both RF and GBM are properly tuned on training sets.

Kindly help.

**Comment by abhikjha** [comment]
In details this is the problem, when I am using RandomForest classifier on test set (for a binary classification task), this is the confusion matrix -

array([[5230, 7656],
       [ 683, **3340**]], dtype=int64)
and, when I am using GradientBoostingClassifier on same test data, this is the confusion matrix I am getting -

array([[**11404**,  1482],
       [ 2183,  1840]], dtype=int64)
As it can be seen that RF is predicting class 1 (with 3340) correctly while GBM is predicting class 0 (with 11404) correctly. How can I use both these classifiers in some sort of ensemble method so that I can use RF for Class 1 and GBM for class 0?

**Comment by ageron** [comment]
Hi @abhikjha ,

The `VotingClassifier` is presented in chapter 7 (see the code examples in the corresponding Jupyter notebook).  In general, soft voting works best, but it requires the underlying classifiers to have a `predict_proba()` method.  Since you are using a `RandomForestClassifier` and a `GradientBoostingClassifier`, you are in luck: they both have a `predict_proba()` method.
You may want to try adding more diversity of classifiers to the mix, e.g., an SVC (but you must set `probability=True` to have a `predict_proba()` method), etc.  The less the errors are correlated, the better the `VotingClassifier`'s performance will be.

Hope this helps,
Aurélien

**Comment by abhikjha** [comment]
Thanks Aurélien. I have gone through the code examples for VF; it was helpful. However, VF did not help me much (only slightly better results). I will be using as you suggested SVC as well and will see if it results in better performance.

Meanwhile, is there any other way where we can ensemble two classifiers (lets say RF and SVC) and use RF to predict class 0 cases and SVC to predict class 1 cases? Does it sound like cheating?

**Comment by ageron** [comment]
Hi @abhikjha ,
I missed your last question, sorry about that. It sounds like you trust RF more for class 0, and you trust SVC more for class 1. Perhaps you could train a blender: see the section on blending in chapter 7.
In short, you would split your training set in two sets A and B.  Use set A to train RF and SVC. Then use them to make predictions for set B. This will give you two probabilities for the positive class for each instance in set B. You can treat this as a new 2D training set: train a model (e.g., a logistic classifier) on this 2D training set (using the labels of set B).
To classify a new instance, you must use the RF and SVC classifiers to predict the proba for the positive class, which gives you two numbers (a 2D instance), then you use the logistic classifier to classify this 2D instance, and it gives you the final result.
Hope this makes sense,
Aurélien
